### ISSUE TITLE
UIKit Library

### ISSUE THREAD

**Issue by jcampbell05** [hikari-no-yume/touchHLE#81]
There is potential a library which may allow for us to support more UIKit features from early iOS as I know some early games used UITableView for leaderboards (if they didn't do it in OpenGL)

This library allowed some UIKit to run on the Mac and so may provide some clean implementations of common components.

https://github.com/BigZaphod/Chameleon

**Comment by hikari-no-yume** [comment]
Ah, thanks for telling me about that. (Closing the issue just to avoid clutter.)

I think it's not alone actually? I've heard of at least one other project that recreates UIKit.

The problem with these other projects is that integrating them into touchHLE might be as difficult as writing them from scratch. Isn't this one for OS X only?

**Comment by jcampbell05** [comment]
Yeah there are probably others.

So it is focused for MacOS originally but in theory if you reduced the surface are there is a small portion that would work cross platform.

At it's core UIKit is just a wrapper around Quartz/CoreAnimation and they've just recreated that in Chameleon and found a way to host the hierarchy of CALayer's in AppKit as well as send events like mouse clicks in a way that simulates what UIKit would send to the application.

So in-theory if you did the following:

- Find an open source implementation of the low level graphics APIs such as Quartz in GNUStep https://github.com/gnustep
- Host the rendered UIKit hierarchy to a rendering layer from SDL.
- Convert events from SDL land into something Chameleon can use to simulate a UIKit event.
- Strip out the UIKit APIs not commonly needed by games or have been implemented in a way that they just call out to AppKit (I imagine things like VoiceOver)

If we focused on just the text, image and scrollview. That would provide a good initial surface area for most use cases.
